Add GDPR script

Zlopez opened the pull request:
Add SAR (Subject Access Requests) GDPR (General Data Protection Regulation)
script that could be used to gather user data from database.
This script currently takes data from users and
social_auth_usersocialauth table and saves them in json format.

Signed-off-by: Michal Konečný <mkonecny@redhat.com>

bowlofeggs commented on anitya/db/models.py:
```diff
@@ -948,6 +948,31 @@ def get_id(self):
         """
         return six.text_type(self.id)
 
+    def __json__(self, detailed=False):
```
It looks like ```detailed``` does not do anything (and it is also undocumented).

Zlopez replied:
Copied the method from other class and didn't noticed this parameter. Will remove.

bowlofeggs commented on anitya/sar.py:
```diff
@@ -0,0 +1,77 @@
+from anitya.config import config
+from anitya import db
+
+LOG = logging.getLogger('anitya')
```
In Python, naming a variable in all caps conventionally means that it is a constant. The logger isn't a constant, so I recommend naming it ```_log``` (since it is also meant to be private to this module).

Zlopez replied:
Copied from `anitya_cron.py` script. Will change.

bowlofeggs commented on anitya/sar.py:
```diff
@@ -0,0 +1,77 @@
+
+    with open(FILENAME, 'w') as fd:
+        LOG.info('Dump data to json {}'.format(users_list))
+        json.dump(users_list, fd)
```
Note that the SAR data must be written to stdout as documented here:

https://fedora-infra-docs.readthedocs.io/en/latest/sysadmin-guide/sops/gdpr_sar.html#script

Zlopez replied:
I thought the output is `sar_output_file` environment variable. If it should only print on stdout it is much easier to test.

bowlofeggs commented on anitya/tests/test_sar.py:
```diff
@@ -0,0 +1,137 @@
+        sar.main()
+
+        testdata = open(sar.FILENAME).read()
+        print(testdata)
```
Stray print statement.

Zlopez replied:
Will remove.

bowlofeggs commented on anitya/tests/test_sar.py:
```diff
@@ -0,0 +1,137 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © 2014  Red Hat, Inc.
```
2018

Zlopez replied:
Copied from another file. Didn't noticed. Will change.

jeremycline commented on anitya/sar.py:
```diff
@@ -0,0 +1,75 @@
+
+    if SAR_EMAIL:
+        _log.debug('Find users by e-mail {}'.format(SAR_EMAIL))
+        users = users + session.query(db.User).filter(db.User.email == SAR_EMAIL).all()
```
You can write this as ``db.User.query..filter(db.User.email == SAR_EMAIL).all()`` or even ``db.User.query.filter_by(email=SAR_EMAIL).all()``.

Zlopez replied:
Didn't know you can write it like this. It is more readable. I will change it.

jeremycline commented on anitya/db/models.py:
```diff
@@ -954,6 +954,31 @@ def get_id(self):
         """
         return six.text_type(self.id)
 
+    def __json__(self):
```
I know this is copied from other model, but I do not think the name of this method conveys what it does. It's not a "real" dunder method, nor is it actually returning JSON. ``as_dict`` or ``to_dict`` or something is clearer.

Zlopez replied:
Yeah, you are right.

jeremycline commented on anitya/tests/test_sar.py:
```diff
@@ -0,0 +1,130 @@
+
+        out, err = self.capsys.readouterr()
+
+        self.assertTrue('"username": "' + user.username + '"' in out)
```
Using a format string here would make it easier to read:
```suggestion
        self.assertTrue('"username": "{}"'.format(user.username) in out)
```

Zlopez replied:
I will change this.

Zlopez commented:
@bowlofeggs is this looking good to you now?

bowlofeggs commented on anitya/tests/test_sar.py:
```diff
@@ -0,0 +1,130 @@
+        out, err = self.capsys.readouterr()
+        print(out)
+
+        self.assertTrue('' in out)
```
This will be ```True``` for any ```str```:

```
>>> '' in 'hello'
True
```

I recommend asserting that the output is exactly the JSON string that is expected (would that be ```[]```?)

Zlopez replied:
Thanks for noticing. I will change this to work with json.

bowlofeggs commented on anitya/tests/test_sar.py:
```diff
@@ -0,0 +1,130 @@
+        self.assertTrue('"username": "{}"'.format(user.username) in out)
+        self.assertTrue('"email": "{}"'.format(user.email) in out)
+        self.assertFalse('"username": "{}"'.format(user2.username) in out)
+        self.assertFalse('"email": "{}"'.format(user2.email) in out)
```
Rather than spot checking strings, why not load the output with ```json.loads()``` and asserting that it is exactly the expected structure, including all keys?

Zlopez replied:
Copied this from some other test, but you are right, it will be better to check against expected JSON. I will change this.

bowlofeggs commented on anitya/tests/test_sar.py:
```diff
@@ -0,0 +1,130 @@
+        out, err = self.capsys.readouterr()
+
+        self.assertTrue('"username": "{}"'.format(user.username) in out)
+        self.assertTrue('"email": "{}"'.format(user.email) in out)
```
Similarly here, it would be good to assert that the output is valid JSON and has all the expected keys.

bowlofeggs commented on anitya/sar.py:
```diff
@@ -0,0 +1,74 @@
+_log = logging.getLogger('anitya')
+
+SAR_USERNAME = os.getenv('SAR_USERNAME')
+SAR_EMAIL = os.getenv('SAR_EMAIL')
```
It would be easier to test this if these lines were part of main. Then you can use ```mock``` to set the environment variables and assert that these lines use ```os.getenv()``` correctly.

Zlopez replied:
Yes, you are right. Thanks for noticing.

Zlopez commented:
@bowlofeggs I addressed your changes and did a squash,

Zlopez commented:
Rebased with latest master

bowlofeggs commented on anitya/tests/test_sar.py:
```diff
@@ -0,0 +1,156 @@
+        }]
+
+        with mock.patch('os.getenv') as mock_getenv:
+            mock_getenv.side_effect = mock_getenv_email
```
Why not use [```mock.patch.dict()```](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch.dict) here instead? It's less code for you to maintain and is clearer for readers.

Zlopez replied:
Didn't thought about this. You are right, it will be clearer.

bowlofeggs commented on anitya/tests/test_sar.py:
```diff
@@ -0,0 +1,156 @@
+            }
+        ]
+
+        sar.SAR_USERNAME = 'user'
```
This line doesn't seem needed anymore.

Zlopez replied:
Probably missed it.

Zlopez commented:
I fixed the conflict and addressed the last suggestions.
There are still failing tests, but these will be fixed by https://github.com/release-monitoring/anitya/pull/667

bowlofeggs commented:
Cool, let me know once tests pass.

Zlopez commented:
@bowlofeggs 
Tests are working again. \o/
